Dependency Modernization and Metadata Writing Features

ballaballaballa opened the pull request:
This PR introduces several improvements including dependency updates, EPUB metadata writing capabilities, and source file refactoring.

### Summary

**Dependency Modernization:**
- Migrate from `blackboxprotobuf` to `bbpb` (new PyPI package name, import name unchanged)
- Replace deprecated `appdirs` with actively maintained `platformdirs`
- Add minimum version constraints for all dependencies
- Set `requires-python = ">=3.9"` for Python 3.9-3.13 compatibility
- Make `tomli` conditional (only required for Python < 3.11)

**New Features:**
- Add EPUB metadata writing support for Storytel and Nextory
- Allow metadata writing to be triggered via command-line parameter or config file
- Add configurable output path option

**Refactoring:**
- Refactor Storytel source file for improved maintainability
- Refactor Nextory source file for improved maintainability

**Documentation:**
- Update README with correct config file paths

### Changes by File

- **Dependencies** (`pyproject.toml`, `shell.nix`): Updated package names and added version constraints
- **Config** (`grawlix/config.py`): Migrated to platformdirs, added output path configuration
- **New modules**: `grawlix/epub_metadata.py`, `grawlix/epub_metadata_writers.py`
- **CLI** (`grawlix/__main__.py`, `grawlix/arguments.py`): Added metadata writing options
- **Sources** (`grawlix/sources/storytel.py`, `grawlix/sources/nextory.py`): Refactored and added metadata support
- **Output** (`grawlix/output/__init__.py`): Enhanced with metadata writing capabilities
- **Documentation** (`README.md`): Updated with new features and correct paths

### Testing

All changes maintain backward compatibility while adding new optional features. The migration to `bbpb` and `platformdirs` uses packages that are drop-in replacements with the same APIs.

ballaballaballa committed:
Update README with correct paths for config file

ballaballaballa committed:
Added functionality to write metadata to epub. Only for Storytel for now. Can be triggered with parameter or set in config.
Added option to add output path to config file.

ballaballaballa committed:
feat: add metadata writing support for Nextory and refactor source files

- Fix Nextory SSL certificate error (api.nextory.se -> api.nextory.com)
- Add missing X-OS-INFO header for Nextory API compatibility
- Implement full metadata writing for Nextory (title, authors, translators, series, publisher, ISBN, language, description, release date)
- Remove debug JSON output from Nextory authentication
- Fix exit() call in epub.py that prevented metadata writing for EpubInParts

Refactoring improvements:
- Reorganize functions in nextory.py and storytel.py by execution flow
- Add section comments for better code organization (Authentication, Main download, Book path, Series path)
- Remove duplicate imports (OnlineFile in nextory.py, datetime in storytel.py)
- Remove duplicate header update in storytel.py
- Improve pagination logic in storytel.py with clearer variable names
- Optimize API calls in nextory.py (eliminate redundant call)
- Use list comprehensions for cleaner code

Documentation:
- Update README with Nextory metadata support
- Add supported sources comparison table for metadata fields
- Include Nextory configuration example

ballaballaballa committed:
feat: migrate to bbpb and platformdirs, add dependency version constraints
- Migrate from blackboxprotobuf to bbpb (new PyPI package name)
  - Update pyproject.toml and shell.nix dependencies
  - Import name remains 'blackboxprotobuf' (no code changes)
- Migrate from appdirs to platformdirs
  - Replace deprecated appdirs with actively maintained platformdirs
  - Update grawlix/config.py to use platformdirs.user_config_dir
  - Update dependency declarations in pyproject.toml and shell.nix
- Add minimum version constraints for all dependencies
  - Set requires-python = ">=3.9" (required by httpx and importlib.resources.files())
  - Add Python version classifiers (3.9-3.13)
  - Specify minimum versions: beautifulsoup4>=4.9.0, bbpb>=1.0.0,
    EbookLib>=0.17, httpx>=0.23.0, importlib-resources>=5.0,
    lxml>=4.6.0, platformdirs>=3.0.0, pycryptodome>=3.10.0, rich>=10.0.0
  - Make tomli conditional: only required for Python < 3.11
- Update shell.nix with correct SHA256 hash for bbpb 1.4.2
These changes ensure all dependencies meet minimum version requirements
for Python 3.9+ compatibility and replace deprecated packages with
actively maintained alternatives.

ballaballaballa committed:
feat: fix Nextory series null handling and add PDF book support

- Fix _extract_series_name to handle null series values
- Add PDF book support for Nextory (served via epub endpoint)
- Convert PDF-in-epub to merged PDF after download
- Add pypdf dependency for PDF merging
- Add Pdf output format (unused but available)
- Update metadata transformer to handle both epub and pdf formats

ballaballaballa committed:
chore: bump version to 0.2.2

ballaballaballa committed:
fix: detect PDF-in-epub by content instead of API format type

jo1gi commented on grawlix/epub_metadata.py:
```diff
@@ -0,0 +1,303 @@
+"""
```
I think it makes sense for this file to be in the output directory

jo1gi commented on grawlix/pdf_converter.py:
```diff
@@ -0,0 +1,68 @@
+"""
```
This file should be in the output directory

jo1gi reviewed (commented):
Thank you for your contribution and sorry for answering this late.

ballaballaballa commented:
> Thank you for your contribution and sorry for answering this late.

No worries! I agree with your comments. I'm away for a couple of weeks, so can take care of it when I'm back if you don't want to fix it before that.

ballaballaballa committed:
refactor: move output-related modules to output folder

- Move pdf_converter.py to output/
- Move epub_metadata.py and epub_metadata_writers.py to output/metadata/
- Update imports in __main__.py

ballaballaballa committed:
refactor: simplify EPUB metadata architecture

- Remove source_data/transformer pattern in favor of direct Metadata population
- Extend Metadata with new fields: original_title, translators, category, tags
- Rename identifier field to isbn for clarity
- Delete epub_metadata_writers.py (transformers no longer needed)
- Update epub_metadata.py to accept Metadata object instead of dict
- Update Storytel and Nextory sources to populate extended metadata

ballaballaballa committed:
fix: resolve compatibility issues from upstream rebase

- Fix f-string quote syntax in ereolen.py for Python < 3.12
- Fix Storytel login signature to match new url parameter
- Fix Nextory API field names (identifier vs isbn confusion)

ballaballaballa commented:
I've made the changes now, so please review again when possible.

ballaballaballa committed:
feat: add fixed-layout support for split EPUB sources

- Add rendition properties to Metadata (layout, spread, orientation)
- Extract and preserve rendition properties when merging EPUB parts
- Fix viewport meta tags for fixed-layout pages
- Use EpubItem instead of EpubHtml to preserve original content

Fix EPUB validation errors in merged output:
- Skip directory entries, mimetype, and META-INF from manifest
- Exclude nav/toc from spine for fixed-layout books

Improve CSS merging across parts:
- Keep longer version of duplicate selectors (more complete rules)
- Return None from _get_css_rule_key for invalid font-faces

Add cover detection fallback:
- Detect cover from largest image on first page when OPF lacks cover info
- Optimize lookup with dict instead of nested loops

Clean up redundant imports in epub_metadata.py

ballaballaballa committed:
fix: correct series/list downloads for Storytel

download_list_details used the URL's storefront locale segment (e.g.
"se") as the API's includeLanguages param, but Storytel expects real
language codes (e.g. "sv"), causing zero matches and a 404 for any
series/author/narrator list. Also fetch all formats instead of
pre-filtering to ebook server-side, and use dict.get() for
nextPageToken so pagination doesn't crash when the API omits the key
on the last page.

ballaballaballa committed:
bump nextory version

ballaballaballa committed:
fix: handle missing/audiobook-only books gracefully in Storytel series downloads

Storytel's ebook consumable endpoint returns no Location header when a
book has no ebook (e.g. audiobook-only or not yet released), which
crashed with a raw KeyError. Now raises DataNotFound instead.

download_series only caught AccessDenied, so a DataNotFound from one
book aborted the whole series download with no indication of which
book failed. Now both are caught and skipped per-book.

Skip messages (already exists, not found, access denied) were printed
above the live progress display, disconnected from the book's row.
They're now shown inline in place of that book's progress bar.

ballaballaballa committed:
fix: correct series detection and skip audiobook-only books for Nextory

download() checked for "serier" (Danish/Norwegian) in the raw URL, but
Nextory's actual series URLs use the English "/series/" path segment,
so series links fell through to _download_book, which crashed on the
series id with KeyError: 'formats'. Now checks path segments for both.

Also removed a duplicate, dead download() method (the file defined it
twice with identical bodies), and changed _find_format to raise
DataNotFound instead of InvalidUrl when a book has no epub/pdf format
(e.g. audiobook-only), so download_series's existing skip-and-continue
handling covers it instead of aborting the whole series.
